ref: snippet revision plus global cleanup

Julien00859 opened the pull request:
Add the missing snippet revision feature, users can duplicate existing
snippets to edit them. The new snippet links its parent so it is easy to
compare them.

The `GET /` URL has been updated, it takes two optionnal query
parameters: (i) `lang` the language by default selected, it override the
config default. (ii) `parentid` the snippet to copy the code from.

The `POST /new` URL has been updated, it takes a form input `parentid`,
the original snippet the new one is a revision.

Additionnaly to the above feature, the source code have been scanned and
various minor changes have been made. In any order:

* The `newform-button.js` file has been renamed `newform.js`
* The `newform.html` <form> had a `post-snippet` class, it is now an id
* The various icons (save, new, dup, prev) share a common theme
* There is a new `RTDBIN_IDENTSIZE` option that control the length of
  the random snippet identifier. By default 6.
* The form `code` input is now utf-8 encoded (was latin-1)
* The backend form validation is stricter
* The `snippet_id` variable has been renamed `snippetid`
* The redis hset keys were the facto bytestring, it is now explicit
* There is a new function to ease snippet creating in the tests.

While they are not needed by themselves, it is a bit of house keeping
that is sure to be welcome by new contributers.

Co-authored-by: Mestery <48163546+Mesteery@users.noreply.github.com>
Co-authored-by: wyx0-xyz <wyxo.bs09@gmail.com>

Closes #10

Julien00859 commented on bin/controller.py:
```diff
@@ -53,16 +53,16 @@ def post_new():
         raise bt.HTTPError(400, str(exc))
 
-    snippet = models.Snippet.create(code, max(maxusage, 0), lifetime, parentid)
+    snippet = models.Snippet.create(code, ext, max(maxusage, 0), lifetime, parentid)
```
Je suis contre qu'on sauvegarde le langage ou l'extension sur le modèle. Le modèle actuelle veut qu'on génère et partage des url avec une extension relative à la coloration syntaxique à privilégier. L'information est également présente dans le formulaire.

Selon moi, si tu veux forcer qu'une révision aie la même extension que son parent, tu retires le menu de sélection de la langue et tu rajoutes un `<input type=hidden name=lang value={{lang}}>` info que tu peux ajouter dans le formulaire.

Mesteery replied:
D'accord, mais d'ou viendrait `lang` ? Faut bien qu'il soit stocker quelque part

Mesteery replied:
Ce que j'avais fait au début c'était de mettre le langage du parent comme langage par défaut sélectionnée.

Julien00859 commented on bin/controller.py:
```diff
@@ -53,16 +53,16 @@ def post_new():
             maxusage = int(forms.get('maxusage') or maxusage)
             lifetime = Time(forms.get('lifetime') or lifetime)
             parentid = forms.get('parentid', '').encode('ascii')
+            ext = models.Snippet.get_by_id(parentid).ext if parentid else parse_extension(forms.get('lang')) or ext
```
Trop compliqué, voir le commentaire ci-dessous, il est mieux de ne pas dupliquer la source de la langue

Julien00859 commented on bin/views/newform.html:
```diff
@@ -28,12 +28,14 @@
                     <option value="10m">
                 </datalist>
 
+                % if not parentid:
```
Quid de si je veux faire une révision pour montrer une implémentation dans un autre langage ?

Julien00859 commented on bin/models.py:
```diff
@@ -11,15 +11,15 @@
 
 class Snippet:
-    def __init__(self, ident, code, ext, views_left, parentid):
+    def __init__(self, ident, code, ext, views_left, parentid=''):
```
Quelle est cette atrocité ?

Julien00859 commented on tests/test_controller.py:
```diff
@@ -10,9 +10,9 @@
-snippet_htmlxss = Snippet(ident='htmlxss', code='<script>alert("XSS");</script>', views_left=float('+inf'))
+snippet_lipsum = Snippet(ident='lipsum', code="Lipsum", ext='txt', views_left=float('+inf'))
+snippet_python = Snippet(ident='egg', ext='py', code='print("Hello world")', views_left=float('+inf'))
+snippet_htmlxss = Snippet(ident='htmlxss', ext='html', code='<script>alert("XSS");</script>', views_left=float('+inf'))
```
cf plus haut

Julien00859 committed:
ref: snippet revision plus global cleanup

Add the missing snippet revision feature, users can duplicate existing
snippets to edit them. The new snippet links its parent so it is easy to
compare them.

The `GET /` URL has been updated, it takes two optionnal query
parameters: (i) `lang` the language by default selected, it override the
config default. (ii) `parentid` the snippet to copy the code from.

The `POST /new` URL has been updated, it takes a form input `parentid`,
the original snippet the new one is a revision.

Additionnaly to the above feature, the source code have been scanned and
various minor changes have been made. In any order:

* The `newform-button.js` file has been renamed `newform.js`
* The `newform.html` <form> had a `post-snippet` class, it is now an id
* The various icons (save, new, dup, prev) share a common theme
* There is a new `RTDBIN_IDENTSIZE` option that control the length of
  the random snippet identifier. By default 6.
* The form `code` input is now utf-8 encoded (was latin-1)
* The backend form validation is stricter
* The `snippet_id` variable has been renamed `snippetid`
* The redis hset keys were the facto bytestring, it is now explicit
* There is a new function to ease snippet creating in the tests.

While they are not needed by themselves, it is a bit of house keeping
that is sure to be welcome by new contributers.

Co-authored-by: Mestery <48163546+Mesteery@users.noreply.github.com>
Co-authored-by: wyx0-xyz <wyxo.bs09@gmail.com>

Closes #10

Mesteery commented on bin/assets/scripts/newform.js:
```diff
@@ -6,15 +6,13 @@ code.addEventListener('keydown', (event) => {
-
-    const pos = code.selectionStart;
-    const content = code.value;
+    const { value: content, selectionStart, selectionEnd } = code;
```
```suggestion
    const { value, selectionStart, selectionEnd } = code;
```
J'ai été con là dessus, il n'y aucun intérêt à renommer `value` en `content`.

Mesteery commented on bin/assets/scripts/newform.js:
```diff
@@ -6,15 +6,13 @@ code.addEventListener('keydown', (event) => {
 
     // inserts tab at the position of the caret
-    code.value = content.substring(0, pos) + '\t' + content.substring(pos);
+    code.value = `${content.slice(0, selectionStart)}\t${content.slice(selectionEnd)}`;
```
```suggestion
    code.value = `${value.slice(0, selectionStart)}\t${value.slice(selectionEnd)}`;
```
*suite.*
(j'ai pas vu faire en une seule suggestion, vu que je suis sur téléphone.)

Mesteery commented on bin/views/highlight.html:
```diff
@@ -7,5 +7,24 @@
+        <div class="controls">
+            % if parentid:
+            <a class="control" title="Voir la révision précédente" href="/{{parentid}}.{{ext}}">
+                <!-- missing icon -->
```
Que fait-on pour cette icône manquante ?
```suggestion
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
                    <path d="M9 2a2 2 0 00-2 2v8a2 2 0 002 2h6a2 2 0 002-2V6.414A2 2 0 0016.414 5L14 2.586A2 2 0 0012.586 2H9z" />
                    <path d="M3 8a2 2 0 012-2v10h8a2 2 0 01-2 2H5a2 2 0 01-2-2V8z" />
                </svg>
```
Mine de rien, cette icone est plutôt approprié je trouve

fusetim replied:
C'est clairement un "Copier" comme tu l'as dit, ailleurs, je ne me vois pas l'utiliser pour voir la révision d'avant.
> Je vois plus ca comme un "Copier", en tout cas ca fait très bizarre de se dire que ca représente un "Modifier"

fusetim replied:
Quelque chose comme ça peut être ? 

![Frame 1(1)](https://user-images.githubusercontent.com/62183512/106359966-41d76e80-6316-11eb-8c0a-8161008d0a21.png)

```suggestion
                <svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" fill="currentColor">
                    <path d="M4.59 2.59C4.96 2.21 5.47 2 6 2H10.59C11.12 2 11.63 2.21 12 2.59L15.41 6C15.79 6.37 16 6.88 16 7.41V16C16 16.53 15.79 17.04 15.41 17.41 15.04 17.79 14.53 18 14 18H6C5.47 18 4.96 17.79 4.59 17.41 4.21 17.04 4 16.53 4 16V4C4 3.47 4.21 2.96 4.59 2.59ZM9.86 7.29C9.72 7.15 9.52 7.07 9.33 7.07 9.13 7.07 8.94 7.15 8.8 7.29 8.65 7.43 8.58 7.63 8.58 7.82V11.32C8.58 11.47 8.62 11.62 8.7 11.74 8.79 11.87 8.91 11.96 9.05 12.02L11.55 13.02C11.73 13.09 11.93 13.08 12.11 13 12.29 12.92 12.43 12.78 12.5 12.59 12.57 12.41 12.57 12.21 12.5 12.03 12.42 11.85 12.28 11.71 12.1 11.63L10.08 10.82V7.82C10.08 7.63 10 7.43 9.86 7.29Z"/>
                 </svg>
```

Mesteery replied:
Ca a l'air magnifique, mais a mon avis, faudrait juste que "les flèches" soient plus "large", parce que ca fait un peu fin.
![image](https://user-images.githubusercontent.com/48163546/106360409-92e86200-6318-11eb-9c3f-3884929026e1.png)
Un peu comme la largeur du "plus", mais moins

fusetim replied:
Quelque chose comme ça :

![Frame 4(1)](https://user-images.githubusercontent.com/62183512/106361118-8239eb00-631c-11eb-8008-eb49454ed9f3.png)

Size ref: 
![image](https://user-images.githubusercontent.com/62183512/106361160-b9100100-631c-11eb-96e1-188b3296f607.png)


```suggestion
                <svg viewBox="0 0 20 20" xmlns="http://www.w3.org/2000/svg" fill="currentColor">
                    <path d="M4.6 2.6C5 2.2 5.5 2 6 2H10.6C11.1 2 11.6 2.2 12 2.6L15.4 6C15.8 6.4 16 6.9 16 7.4V16C16 16.5 15.8 17 15.4 17.4 15 17.8 14.5 18 14 18H6C5.5 18 5 17.8 4.6 17.4 4.2 17 4 16.5 4 16V4C4 3.5 4.2 3 4.6 2.6ZM9.7 7.3C9.5 7.1 9.3 7 9 7 8.7 7 8.5 7.1 8.3 7.3 8.1 7.5 8 7.7 8 8V12.7C8 12.9 8.1 13.1 8.2 13.2 8.3 13.4 8.4 13.5 8.6 13.6L12 14.9C12.2 15 12.5 15 12.7 14.9 12.9 14.8 13.1 14.6 13.2 14.4 13.3 14.1 13.3 13.8 13.2 13.6 13.1 13.4 12.9 13.2 12.7 13.1L10 12V8C10 7.7 9.9 7.5 9.7 7.3Z"/>
                </svg>
```

Mesteery commented on bin/views/highlight.html:
```diff
@@ -7,5 +7,24 @@
+                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
+                  <path d="M9 2a2 2 0 00-2 2v8a2 2 0 002 2h6a2 2 0 002-2V6.414A2 2 0 0016.414 5L14 2.586A2 2 0 0012.586 2H9z" />
+                  <path d="M3 8a2 2 0 012-2v10h8a2 2 0 01-2 2H5a2 2 0 01-2-2V8z" />
+                </svg>
```
Je vois plus ca comme un "Copier", en tout cas ca fait très bizarre de se dire que ca représente un "Modifier"
![image](https://user-images.githubusercontent.com/48163546/106358668-e86b4180-630d-11eb-97e8-6b4af29366cc.png)

Pourquoi ne pas prendre un des deux ?
![image](https://user-images.githubusercontent.com/48163546/106358725-36804500-630e-11eb-9a99-bb7616d1bf80.png)

Julien00859 replied:
Normale... ça copie justement, je trouve l'icone actuelle parfaitement adaptée, celle permet de faire une copie modifiée

Mesteery replied:
Ca me fait quand même bizarre

Julien00859 committed:
Update bin/views/highlight.html

Co-authored-by: FuseTim <62183512+fusetim@users.noreply.github.com>

Mesteery reviewed (approved):
C'est parfait, mais il y a quelques petits problèmes d'indentation dans le HTML (autour des SVG je crois).

fusetim commented on bin/views/newform.html:
```diff
@@ -30,14 +31,17 @@
                     <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
-                        <path d="M17.414 2.586a2 2 0 00-2.828 0L7 10.172V13h2.828l7.586-7.586a2 2 0 000-2.828z" />
-                        <path fill-rule="evenodd" d="M2 6a2 2 0 012-2h4a1 1 0 010 2H4v10h10v-4a1 1 0 112 0v4a2 2 0 01-2 2H4a2 2 0 01-2-2V6z" clip-rule="evenodd" />
+                      <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm5 6a1 1 0 10-2 0v3.586l-1.293-1.293a1 1 0 10-1.414 1.414l3 3a1 1 0 001.414 0l3-3a1 1 0 00-1.414-1.414L11 11.586V8z" clip-rule="evenodd" />
```
```suggestion
                        <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm5 6a1 1 0 10-2 0v3.586l-1.293-1.293a1 1 0 10-1.414 1.414l3 3a1 1 0 001.414 0l3-3a1 1 0 00-1.414-1.414L11 11.586V8z" clip-rule="evenodd" />
```

fusetim reviewed (approved):
C'est tout bon, éventuellement rajouter l'indentation manquante mais sinon tout marche. :+1:

Mesteery commented on bin/views/highlight.html:
```diff
@@ -7,5 +7,25 @@
+            <a class="control" title="Nouveau" href="/">
+                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
+                  <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm5 6a1 1 0 10-2 0v2H7a1 1 0 100 2h2v2a1 1 0 102 0v-2h2a1 1 0 100-2h-2V8z" clip-rule="evenodd" />
+                </svg>
```
```suggestion
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
                    <path d="M9 2a2 2 0 00-2 2v8a2 2 0 002 2h6a2 2 0 002-2V6.414A2 2 0 0016.414 5L14 2.586A2 2 0 0012.586 2H9z" />
                    <path d="M3 8a2 2 0 012-2v10h8a2 2 0 01-2 2H5a2 2 0 01-2-2V8z" />
                </svg>
            </a>
            <a class="control" title="Nouveau" href="/">
                <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 20 20" fill="currentColor">
                    <path fill-rule="evenodd" d="M6 2a2 2 0 00-2 2v12a2 2 0 002 2h8a2 2 0 002-2V7.414A2 2 0 0015.414 6L12 2.586A2 2 0 0010.586 2H6zm5 6a1 1 0 10-2 0v2H7a1 1 0 100 2h2v2a1 1 0 102 0v-2h2a1 1 0 100-2h-2V8z" clip-rule="evenodd" />
                </svg>
```

Julien00859 committed:
Update bin/views/highlight.html

Co-authored-by: Mestery <48163546+Mesteery@users.noreply.github.com>
